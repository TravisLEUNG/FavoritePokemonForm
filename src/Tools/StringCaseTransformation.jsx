export function _snake2Label( str ){
    str +='';
    str = str.split('_');
    for(var i=0;i<str.length;i++){ 
        str[i] = str[i].slice(0,1).toUpperCase() + str[i].slice(1,str[i].length);
    }
    return str.join(' ');
}

export function _hyphen2Label( str ){
    str +='';
    str = str.split('-');
    for(var i=0;i<str.length;i++){ 
        str[i] = str[i].slice(0,1).toUpperCase() + str[i].slice(1,str[i].length);
    }
    return str.join(' ');
}

export function _camel2Label( str ){
    return str.slice(0,1).toUpperCase()+str.slice(1,str.length).replace(/[A-Z]/g, letter => ` ${letter.toLowerCase()}`);
}