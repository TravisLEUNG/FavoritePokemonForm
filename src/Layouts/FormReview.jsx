import _ from "lodash";
import { _hyphen2Label, _camel2Label } from "../Tools/StringCaseTransformation";
import Table from "react-bootstrap/Table";
import ButtonField from "../Components/ButtonField";

export default function FormReview(props) {
  let tableRow = [];
  for (const [field, value] of Object.entries(props.data)) {
    if (_.isObject(value)) {
      tableRow.push(
        <tr key={field}>
          <th>{_camel2Label(field)}</th>
          <td data-testid={`data-${field}`}>{_hyphen2Label(value.value || "")}</td>
        </tr>
      );
    } else {
      tableRow.push(
        <tr key={field}>
          <th>{_camel2Label(field)}</th>
          <td data-testid={`data-${field}`}>{value || ""}</td>
        </tr>
      );
    }
  }
  return (
    <div className="form-review">
      <h3>Review</h3>
      <p>Please double check the information.</p>
      <Table className="form-review_content">
        <tbody>{tableRow}</tbody>
      </Table>
      <div className="d-grid gap-3 form-review_button">
        <ButtonField
          label="Back"
          handleClick={() => props.handleViewChange("Content")}
          isOutline
        />
        <ButtonField isSubmit label="Submit" handleClick={() => props.handleSubmit("Success")}/>
      </div>
    </div>
  );
}
