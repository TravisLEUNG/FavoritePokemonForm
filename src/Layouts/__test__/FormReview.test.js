import { render, screen, cleanup } from "@testing-library/react";
import user from "@testing-library/user-event"
import renderer from "react-test-renderer";
import { _hyphen2Label } from "../../Tools/StringCaseTransformation";

import FormReview from "../FormReview";

const fakeData = {
  FirstName: "",
  LastName: "",
  PhoneNumber: "",
  EmailAddress: "",
  FavoritePokemon: {},
};
const fakeCompletedData = {
  FirstName: "Ultimate",
  LastName: "Hokon",
  PhoneNumber: "+19876543210",
  EmailAddress: "ultimate.hokon@gmail.com",
  FavoritePokemon: {
    label: "Venusaur",
    url: "https://pokeapi.co/api/v2/pokemon/3/",
    value: "venusaur",
  },
};
const fakeEvent = () => {};

describe("FormContent", () => {
  const onViewChange = jest.fn();
  const onSubmit = jest.fn();

  afterEach(() => {
    cleanup();
  });

  it("should render non-completed form data", () => {
    render(<FormReview data={fakeData} handleViewChange={fakeEvent} handleSubmit={fakeEvent}/>);

    for (const field in fakeData) {
      const fieldElement = screen.getByTestId(`data-${field}`);
      expect(fieldElement).toBeInTheDocument();
      expect(fieldElement).toBeEmptyDOMElement();
    }
  });

  it("should render completed form data", () => {
    const expectedValue = {
      ...fakeCompletedData,
      FavoritePokemon: _hyphen2Label(
        fakeCompletedData["FavoritePokemon"].value
      ),
    };

    render(
      <FormReview data={fakeCompletedData} handleViewChange={fakeEvent} handleSubmit={fakeEvent}/>
    );
    for (const field in expectedValue) {
      const fieldElement = screen.queryByTestId(`data-${field}`);
      expect(fieldElement).toBeInTheDocument();
      expect(fieldElement).toHaveTextContent(expectedValue[field]);
    }
  });

  it("Submit when click button", () => {
    render(<FormReview data={fakeData} handleViewChange={onViewChange} handleSubmit={onSubmit}/>);
    clickSubmitButton();
    expect(onViewChange).toHaveBeenCalledTimes(0);
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });

  it("Back when click button", () => {
    render(<FormReview data={fakeData} handleViewChange={onViewChange}  handleSubmit={onSubmit}/>);
    clickBackButton();
    expect(onViewChange).toHaveBeenCalledTimes(1);
    expect(onSubmit).toHaveBeenCalledTimes(0);
  });

  test("Match snapshot", () => {
    const tree = renderer
      .create(<FormReview data={fakeCompletedData} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

function clickBackButton() {
  return user.click(screen.getByRole('button', { name: /Back/i }));
}
function clickSubmitButton() {
  return user.click(screen.getByRole('button', { name: /Submit/i }));
}