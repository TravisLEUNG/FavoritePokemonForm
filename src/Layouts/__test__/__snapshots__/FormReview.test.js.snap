// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContent Match snapshot 1`] = `
<div
  className="form-review"
>
  <h3>
    Review
  </h3>
  <p>
    Please double check the information.
  </p>
  <table
    className="form-review_content table"
  >
    <tbody>
      <tr>
        <th>
          First name
        </th>
        <td
          data-testid="data-FirstName"
        >
          Ultimate
        </td>
      </tr>
      <tr>
        <th>
          Last name
        </th>
        <td
          data-testid="data-LastName"
        >
          Hokon
        </td>
      </tr>
      <tr>
        <th>
          Phone number
        </th>
        <td
          data-testid="data-PhoneNumber"
        >
          +19876543210
        </td>
      </tr>
      <tr>
        <th>
          Email address
        </th>
        <td
          data-testid="data-EmailAddress"
        >
          ultimate.hokon@gmail.com
        </td>
      </tr>
      <tr>
        <th>
          Favorite pokemon
        </th>
        <td
          data-testid="data-FavoritePokemon"
        >
          Venusaur
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="d-grid gap-3 form-review_button"
  >
    <button
      className="btn btn-outline-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Back
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</div>
`;
